builtin/logical/totp/path_roles.go

```diff
@@ -0,0 +1,342 @@
+	}
+
+	// Enforce input value requirements
+	if period <= 0 {
+		return logical.ErrorResponse("The period value must be greater than zero."), nil
+	}
+
+	if skew < 0 {
+		return logical.ErrorResponse("The skew value must be greater than zero."), nil
+	}
+
+	if qr_size <= 0 {
+		return logical.ErrorResponse("The qr_size value must be greater than zero."), nil
+	}
+
+	if key_size <= 0 {
+		return logical.ErrorResponse("The key_size value must be greater than zero."), nil
+	}
+
+	// If the key is generated, Account Name and Issuer are required.
+	if generate {
+		if account_name == "" {
+			return logical.ErrorResponse("The account_name value is required for generated keys."), nil
+		}
+
+		if issuer == "" {
+			return logical.ErrorResponse("The issuer value is required for generated keys."), nil
+		}
+	}
+
+	// Period, Skew and Key Size need to be unsigned ints
+	uint_period := uint(period)
+	uint_skew := uint(skew)
+	uint_key_size := uint(key_size)
+
+	url := ""
+
+	switch generate {
+	case true:
+		// Generate a new key
+		key_object, err := totplib.Generate(totplib.GenerateOpts{
+			Issuer:      issuer,
+			AccountName: account_name,
+			Period:      uint_period,
+			Digits:      role_digits,
+			Algorithm:   role_algorithm,
+			SecretSize:  uint_key_size,
+		})
+
+		if err != nil {
+			return logical.ErrorResponse("An error occured while generating a key."), nil
+		}
+
+		url = key_object.String()
+		key_string = key_object.Secret()
+	case false:
+		if key_string == "" {
+			return logical.ErrorResponse("The key value is required."), nil
+		}
+
+		_, err := base32.StdEncoding.DecodeString(key_string)
+
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf(
+				"Invalid key value: %s", err)), nil
+		}
+	}
+
+	// Store it
+	entry, err := logical.StorageEntryJSON("key/"+name, &roleEntry{
+		Key:         key_string,
+		Issuer:      issuer,
+		AccountName: account_name,
+		Period:      uint_period,
+		Algorithm:   role_algorithm,
+		Digits:      role_digits,
+		Skew:        uint_skew,
+		QRSize:      qr_size,
+		URL:         url,
+		Generate:    generate,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if err := req.Storage.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+type roleEntry struct {
+	Key         string           `json:"key" mapstructure:"key" structs:"key"`
+	Issuer      string           `json:"issuer" mapstructure:"issuer" structs:"issuer"`
+	AccountName string           `json:"account_name" mapstructure:"account_name" structs:"account_name"`
+	Period      uint             `json:"period" mapstructure:"period" structs:"period"`
+	Algorithm   otplib.Algorithm `json:"algorithm" mapstructure:"algorithm" structs:"algorithm"`
+	Digits      otplib.Digits    `json:"digits" mapstructure:"digits" structs:"digits"`
+	Skew        uint             `json:"skew" mapstructure:"skew" structs:"skew"`
+	QRSize      int              `json:"qr_size" mapstructure:"qr_size" structs:"qr_size"`
```

Comment:
The library we are using to generate the barcode requires an int as input for width and height.